### ISSUE TITLE
Migrate tensor backend from tch (libtorch FFI) to Burn for multi-backend GPU support

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/thrust#65]
## Background

Thrust currently uses `tch = 0.22` (Rust bindings to libtorch) as its tensor backend, gated behind the `training` feature. This is the right choice today — libtorch is mature, the autodiff is well-tested, and the WASM-side pure-Rust inference path (`src/policy/{inference, universal_inference}.rs`) deliberately avoids tch — but it has structural drawbacks:

- **Heavyweight C++ FFI**. Installing libtorch is a real onboarding cost (see `docs/LIBTORCH_SETUP.md`, `docs/GPU_SETUP.md`, `docs/GPU_TRAINING_GUIDE.md`, `scripts/download-libtorch.sh`, `scripts/setup-libtorch.sh`).
- **Single-vendor GPU story**. libtorch CUDA paths work; ROCm/Metal/Vulkan/WebGPU portability is not a first-class concern of the Rust bindings.
- **No native tensor IR**. tch defers to PyTorch's eager-mode runtime; we can't lower kernels through a Rust-native IR for cross-platform / WASM-side training.

Maintainer direction (per #64): **before v1.0, move to a Burn-based backend** ([burn.dev](https://burn.dev)). Sister project `../geode-fem` is already Burn-based.

## Why Burn

Same reasons `../geode-fem`'s README cites:

- Multi-backend GPU without per-backend kernels (CUDA, ROCm, Metal, Vulkan, WebGPU)
- Native Rust autodiff
- Tensor IR usable from WASM (would let us unify the training-side and inference-side stacks)
- Active open-source development; no FFI to a vendor runtime
- Composable with Rust-native ML primitives (no Python in the loop)

## Sub-issues / phases

This umbrella issue tracks the decomposition into six phases. Each phase is a self-contained, single-PR Builder task with its own acceptance criteria. Phases proceed strictly in order — each is blocked by the previous one.

- **Phase 1 — Scout:** #78 (`Burn migration phase 1: scout — port simple_bandit trainer behind training-burn feature`). The architectural-fit spike. Adds a parallel `training-burn` feature flag and ports `train_simple_bandit` end-to-end to validate that Burn supports the patterns we need before committing to the full migration. **Deliverable: a friction-points report that re-sizes phases 2–4.**
- **Phase 2 — Tensor/Optimizer surface:** #79 (`Burn migration phase 2: port src/buffer/ and src/train/ optimizer/tensor surface to Burn (parallel to tch)`). Migrates `RolloutBatch`/`ReplayBatch::to_tensors`, the optimizer abstraction, and the device helper to support both backends in parallel. Algorithm bodies and policies untouched.
- **Phase 3 — Algorithms:** #80 (`Burn migration phase 3: port PPO + DQN loss/update to compile under both backends`). Ports PPO and DQN loss math and the train-step logic to backend-generic form. Includes numerical-parity unit tests at 1e-4 absolute tolerance.
- **Phase 4 — Policy networks:** #81 (`Burn migration phase 4: port MlpPolicy, MultiDiscreteMlpPolicy, SnakeCnn, QNetwork to Burn Module`). The four policy networks become `burn::module::Module` impls. Multi-agent learner/joint paths migrate with them.
- **Phase 5 — Drop tch:** #82 (`Burn migration phase 5: remove tch path, rename training-burn → training, drop libtorch docs/scripts`). The irreversible deletion phase. Maintainer approval required before merge.
- **Phase 6 — WASM unification (OPTIONAL):** #83 (`Burn migration phase 6 (optional): unify WASM inference under Burn WebGPU backend`). Replaces the hand-rolled pure-Rust inference path with Burn WebGPU. Hard bundle-size and latency gates; abandonable if Burn's WASM story isn't ready.

## Related pre-migration cleanups (independent, recommended-but-not-required)

These are tracked separately because they improve the tch-era code on their own merits and shrink the migration PRs if they land first:

- #84: Extract tch tensor-construction boilerplate from example trainers into a helper.
- #85: Prune legacy CartPole and Snake example trainer variants (deletion-heavy maintenance pass).

## Status

- [ ] Phase 1 (#78) — scout
- [ ] Phase 2 (#79) — tensor/optimizer surface
- [ ] Phase 3 (#80) — PPO + DQN algorithms
- [ ] Phase 4 (#81) — policy networks
- [ ] Phase 5 (#82) — drop tch
- [ ] Phase 6 (#83) — optional WASM unification

## Risks and known unknowns

- **Burn's autodiff and CUDA performance vs. libtorch on real training workloads.** The scout phase needs to validate that we don't lose meaningful throughput.
- **Burn API churn.** Burn is itself pre-1.0 (`burn = \"0.21\"` in geode-fem). A migration to a pre-1.0 dependency carries the usual risk; mitigated by Burn having more active maintenance momentum than the alternatives.
- **GPU determinism guarantees.** PPO and DQN reproducibility under a different RNG / autodiff implementation needs verification.
- **Optimizer-type impedance mismatch.** Burn's optimizer binds the model into its type (`Optim<Model, B>`); tch's holds a `VarStore` reference. This is the single most likely source of API friction during phase 2.
- **The WASM inference path may not need to migrate.** It's already pure-Rust by design. The unification in phase 6 is a \"could\", not a \"must\".

## Out of scope

- Adding new algorithms in the migration PRs. Algorithm work goes on the existing backlog (#58, #59 for DQN improvements; centralized-critic MAPPO in #63; SAC in #61's downstream follow-up).
- Re-architecting the multi-agent infra. `src/multi_agent/` should migrate with as little structural change as possible; behavioral changes go in separate issues.

## Acceptance criteria for closing this umbrella issue

All criteria must be true after the merged phases (1 through 5 mandatory; 6 optional):

- [ ] All existing trainer examples (CartPole, Snake single + multi-agent, Pong + self-play, DQN) compile and produce reasonable training curves under Burn.
- [ ] `tch` is fully removed from the workspace.
- [ ] CI runs no longer install libtorch.
- [ ] `Cargo.toml` no longer has the `training` (tch) feature; the Burn path is the default.
- [ ] Docs (`README.md`, `docs/PPO_BEST_PRACTICES.md`, etc.) updated to reflect the new backend.
- [ ] At least one Burn backend other than the default (e.g., WebGPU or wgpu) is verified to train at least one env end-to-end.

## Why this lives in the backlog now

Surfaced as the planned post-v0.1.0 direction during the v0.1.0 release prep in #64. We're publishing the tch-based v0.1.0 to crates.io while making it clear in the README that the backend will change before v1.0.

## Related

- #64 — Release machinery + v0.1.0 cut (the immediate work; this issue is the bigger thing it foreshadows).
- `../geode-fem` — sister project, already Burn-based, can be referenced as a working example of the target stack.

**Comment by rjwalters** [comment]
## Decomposed into phased epic

This umbrella issue is now an epic tracker (label: `loom:epic`). Six phase sub-issues were opened, each sized for a single Builder PR. Two independent pre-migration cleanup proposals were also filed.

### Dependency order

```
#78 (phase 1: scout)
  └─> #79 (phase 2: tensor/optimizer surface)
        └─> #80 (phase 3: PPO + DQN algorithms)
              └─> #81 (phase 4: policy networks)
                    └─> #82 (phase 5: drop tch)
                          └─> #83 (phase 6: WASM unification, OPTIONAL)
```

Each phase issue is labeled `loom:architect` (proposal awaiting human approval) + `loom:epic-phase` (part of this epic). They should not be promoted to `loom:issue` (ready for Builder) one-by-one in order, with the previous phase **merged** before the next phase is approved.

### Independent cleanup proposals

These are NOT phases of the epic. They land independently on the current tch stack and reduce migration PR sizes if they land before phase 4:

- **#84:** Extract tensor-construction boilerplate from example trainers into a helper. (~120 LOC net deletion; shrinks phase 4 measurably.)
- **#85:** Prune legacy CartPole and Snake example trainer variants. (Deletion-heavy; needs maintainer sign-off on the keepers/deletes list.)

Both are labeled `loom:architect` + `tier:maintenance`.

### Sequencing risks worth flagging up-front

1. **Phase 1 friction report is load-bearing.** The acceptance gate for phase 1 is not just \"bandit trains under Burn\" — it's the friction-points report that re-sizes phases 2–4. If that report surfaces blockers (e.g., Burn's optimizer-type ergonomics are worse than expected with the existing `PPOTrainer<P>` generic design), phases 2 and 3 may need to be replanned or split. Don't auto-promote #79 until #78 lands and its report has been reviewed.

2. **Phase 5 is irreversible and should not autopilot.** Once tch is deleted, rollback means resurrecting code from git history. Recommend a 2-week soak between phase 4 merge and phase 5 PR start, including at least one long training run (Pong self-play on alc-2 — see #75) under the Burn-only stack.

3. **Phase 4 has the largest LOC delta** (~1000 LOC target). The multi-agent `src/multi_agent/joint.rs` (1056 LOC) plus the four policy networks plus the canonical example trainers stretch the per-PR sizing. If phase 4 grows beyond ~1500 LOC during implementation, split off \"phase 4b: src/multi_agent migration\" as a follow-up issue.

4. **Phase 6 has hard exit gates.** WASM bundle size or inference latency growing >2x triggers abandonment of phase 6 with a permanent \"two-stack inference is the cheaper option\" decision documented here. The epic can be closed without phase 6 ever landing.

5. **Burn version pinning matters.** `../geode-fem` uses `burn = \"0.21\"`. Phase 1 should pin the same version; bumping Burn mid-migration would be painful. Allow upgrades only at phase boundaries.

### Per architect workflow

This epic itself is NOT labeled `loom:issue` (it's not directly Buildable). Phase issues stay at `loom:architect` until promoted by a human reviewer. Open architect-proposal count was 0 before this batch; it is now 8 (6 phases + 2 cleanups), which exceeds the typical 3-proposal cap — flagging that explicitly so the next Architect iteration knows to stand down on proposals unrelated to this epic until the backlog drains.

**Comment by rjwalters** [comment]
## Phase 2 split into 2a + 2b (post-scout)

The phase 1 scout (#78, merged in PR #87) confirmed friction point #1 from its report: Burn's optimizer ownership model (\`optim.step(lr, module, grads) -> module\`) makes the existing \`PPOTrainer<P> { optimizer: Option<tch::nn::Optimizer> }\` and \`DQNTrainer { optimizer: tch::nn::Optimizer }\` structs incompatible with a feature-gated sum-type variant. The scout's revised estimate for porting the PPO trainer is **600–900 LOC**, not the ~300 a naive port would suggest.

Executing the \"if it's worse than expected, replan as two sub-phases\" branch from the original #79 risk note. Phase 2 is now split into:

### Phase 2a — Buffer/tensor surface (#79, scoped down)

Add \`to_burn_tensors<B: Backend>\` methods to \`ReplayBatch\`, \`PrioritizedBatch\`, \`RolloutBatch\`. Add \`default_burn_device<B>\` parallel to \`default_tch_device\`. **Touches no trainer struct, no optimizer field, no policy.** Mechanical, low risk. Target <300 LOC. Acceptance: existing tch tests still pass; new shape + round-trip tests on the Burn tensor conversions.

### Phase 2b — Optimizer abstraction (#92, NEW)

Introduce a backend-agnostic optimizer abstraction across PPO + DQN trainers. The new issue enumerates three viable patterns (trait-based with associated \`Module\` type; \`PhantomData\` generic on the trainer; two sibling trainer types sharing a trait) and their trade-offs, with the Builder picking one based on prototype findings. Algorithm bodies unchanged at this phase. Target 600–900 LOC. **High risk** — this is the load-bearing surface where the migration's leakiness gets exposed.

### Updated dependency order

\`\`\`
#78 (phase 1: scout — DONE, merged in PR #87)
  ├─> #79 (phase 2a: buffer/tensor surface)        [can land in parallel with 2b]
  └─> #92 (phase 2b: optimizer abstraction)         [blocks phase 3]
        └─> #80 (phase 3: PPO + DQN algorithms)
              └─> #81 (phase 4: policy networks)
                    └─> #82 (phase 5: drop tch)
                          └─> #83 (phase 6: WASM unification, OPTIONAL)
\`\`\`

Phase 2a (#79) and phase 2b (#92) are independent and can be Built in parallel by different sessions. Only #92 blocks phase 3.

### Updated phase 2 totals

| Phase | Original target | Revised target | Notes |
|---|---|---|---|
| 2 (combined, original) | <500 LOC | n/a — split | — |
| 2a (#79, buffer/tensor) | — | <300 LOC | safe half |
| 2b (#92, optimizer)    | — | 600–900 LOC | load-bearing half |

Net: phase 2 in aggregate is ~1.8x the original estimate, matching the scout's prediction of phases 2 and 3 each being 1.5–2x. Phases 3–6 estimates from the scout report are unchanged.

### Status checklist (updated)

- [x] Phase 1 (#78) — scout (merged in PR #87)
- [ ] Phase 2a (#79) — buffer/tensor surface
- [ ] Phase 2b (#92) — optimizer abstraction (NEW)
- [ ] Phase 3 (#80) — PPO + DQN algorithms
- [ ] Phase 4 (#81) — policy networks
- [ ] Phase 5 (#82) — drop tch
- [ ] Phase 6 (#83) — optional WASM unification

### Per architect workflow

#92 is labeled \`loom:architect\` + \`loom:epic-phase\`, **not** \`loom:issue\`. Both 2a and 2b remain proposals awaiting human promotion. Open \`loom:architect\` count after this split: 8 (5 remaining epic phases + phase 2a + phase 2b + the #85 cleanup). The pre-existing cap-exceeded posture from the original epic decomposition still applies — no unrelated proposals until this backlog drains.

**Comment by rjwalters** [comment]
## Sweep audit (post-phases-1-6)

Sweep run on the six phase issues completed; all six PRs (#87, #94, #95, #96, #97, #98, #99) merged. Audit against this issue's acceptance criteria below.

### Status of acceptance criteria

| Criterion | Status | Notes |
|-----------|--------|-------|
| 1. All trainer examples compile + reasonable curves under Burn | ❌ NOT MET | Phase 5 (PR #98) **deleted** every canonical trainer example (cartpole_modern, cartpole_dqn, snake_multi_v2, pong_self_play, pong, bucket_brigade) along with the `multi_agent` module, rather than porting them. Only `train_simple_bandit` survives end-to-end. |
| 2. `tch` fully removed from the workspace | ✅ Met | No `tch::*` imports anywhere in `src/`. |
| 3. CI runs no longer install libtorch | ✅ Met | No libtorch setup in `.github/workflows/`. |
| 4. `Cargo.toml`: training feature is Burn-backed, no tch feature | ✅ Met | `training = ["dep:burn", ...]` is the default. |
| 5. Docs updated to reflect new backend | ⚠️ Mostly met | README and `docs/BURN_BACKENDS.md` reflect the new backend; `docs/ARCHITECTURE_PROPOSAL.md` still has old `use tch::{...}` snippets but that's a historical proposal doc, not user-facing. |
| 6. Non-default Burn backend verified end-to-end | ❌ NOT MET | PR #98 only ran `cargo check --features "training,wgpu"`. No env was actually trained on wgpu. |

### Two unmet criteria → epic stays open

This epic should remain open. PR #98's scope notes openly acknowledge the gap:

> "The `multi_agent` module was tch-coupled and has no Burn port yet… The standalone production trainer examples can be resurrected from git history once the multi-agent layer is back."

That deferral is the work remaining on this epic.

### Follow-up issues filed

- #100 — Port `multi_agent` module to Burn (rebuild deleted infra on top of Burn policies/trainers)
- #101 — Resurrect canonical trainer examples (cartpole, snake_multi, pong_self_play) on Burn stack
- #102 — Validate Burn non-default backend (wgpu or cuda) with an end-to-end training run

#101 is blocked by #100 for the multi-agent trainers; the CartPole + DQN parts of #101 can start independently. #102 is soft-blocked by #101 (more meaningful demo) but can run against `simple_bandit` immediately.

Closing this epic should wait until #100, #101, and #102 all merge — at that point criteria #1 and #6 will both be met.

**Comment by rjwalters** [comment]
## Epic complete — closing

All six acceptance criteria now met after the post-sweep follow-ups landed:

| Criterion | Status | Evidence |
|-----------|--------|----------|
| 1. All trainer examples compile + reasonable curves under Burn | ✅ | PR #104 resurrected cartpole_modern, cartpole_dqn, snake_multi_v2, pong_self_play. PR #105 confirms cartpole_modern reaches reward 175 (vs ~22 random) end-to-end. |
| 2. `tch` fully removed | ✅ | PR #98 — no `tch::*` imports anywhere in `src/`. |
| 3. CI runs no longer install libtorch | ✅ | PR #98 — `.github/workflows/ci.yml` has no libtorch setup. |
| 4. `Cargo.toml`: training feature is Burn-backed | ✅ | PR #98 — `training = ["dep:burn", ...]` is the default. |
| 5. Docs updated | ✅ | PR #98 (README, BURN_BACKENDS.md initial pass) and PR #105 (BURN_BACKENDS.md augmented with wgpu/Metal setup, commands, wall-clock numbers). |
| 6. Non-default Burn backend verified end-to-end | ✅ | PR #105 — `train_simple_bandit` hits 98.3% on wgpu/Metal; `train_cartpole_modern` hits reward 175 on wgpu/Metal (vs reward 22 random baseline). |

### Migration complete

- tch (libtorch FFI) is gone from the workspace.
- Burn is the default and only backend.
- NdArray (CPU) is the default; wgpu (Metal/Vulkan/DX12/WebGPU) and CUDA are opt-in.
- All canonical trainer examples (CartPole PPO + DQN, Snake multi-agent, Pong self-play) are restored and functional on the Burn stack.

Closing this epic. Bucket-brigade trainer resurrection (still gated behind the post-v0.1.0 `env-bucket-brigade` feature) is intentionally deferred and not part of this epic's scope.
